### PR TITLE
Message interactions using Marvin AI

### PR DESCRIPTION
**Pull request by abaddonpuff**
Added interactions with the NPC using Marvin AI. Overall it works well, it verifies the interactions between the npc and the player based on the top values and generates a response by comparing the sentiment of the message values to the npc values and giving points if the right values are shown. Next steps is to provide that answer on the game itself.

### CODE REVIEW

File: mechanics/input_processing.py [hunk 1]
```diff
@@ -1,48 +1,103 @@
...
-    text = """Hello!
-        Hi therwe!
-        Good morning!
-        Good afternoon!"""
-
-    print(process_message(text))
+    with open(str(npc_settings), "r") as npc:
+        npc_personality = json.load(npc)
+    user_input = "Hi"
+    print(generate_response(npc_personality["npc9"]["attributes"], user_input))
```
Comment (bbelderbos) [inline mechanics/input_processing.py:100]: could add a bit of argparse/typer/command line to make it easier to manually test

File: mechanics/input_processing.py [hunk 2]
```diff
@@ -1,48 +1,103 @@
...
+    if sentiment != "neutral" and msg_type == "other":
+        for value in npc_values:
+            if message_values[value] > npc_personality[value]:
+                higher_values += 1
+        if higher_values == 0:
+            sentiment = "boring"
+        if higher_values == 1:
+            sentiment = "neutral"
+        if higher_values == 2:
+            sentiment = "excited"
```
Comment (bbelderbos) [inline mechanics/input_processing.py:89]: can group these constants into an enum

File: mechanics/input_processing.py [hunk 3]
```diff
@@ -1,48 +1,103 @@
...
+
+    return response
+
+
+def evaluate_response(
+    sentiment: str,
+    msg_type: str,
+    npc_personality: dict,
+    npc_values: list,
+    message_values: dict,
```
Comment (bbelderbos) [inline mechanics/input_processing.py:76]: you could further subtype lists and dicts

File: mechanics/input_processing.py [hunk 4]
```diff
@@ -1,48 +1,103 @@
...
+    user_input: str,
+) -> str:
+    higher_values = 0
+    if sentiment != "neutral" and msg_type == "other":
```
Comment (bbelderbos) [inline mechanics/input_processing.py:80]: use constants

File: mechanics/input_processing.py (hunk 2 above)
Comment (bbelderbos) [reply mechanics/input_processing.py:89]: what about > 2 as a case? just "excited" as well?

File: characters/npc/npc_settings.json [hunk 5]
```diff
@@ -1,36 +1,44 @@
...
-        "Humor": 91,
-        "Ambition": 25,
-        "Affection": 33,
-        "Confidence": 61,
-        "Adventurousness": 16,
-        "Empathy": 17
+    "npc9": {
+        "settings": {
+            "name": "npc9",
+            "image_path": "/Users/0xabaddon/new_repos/project_G/images/heart.png",
```
Comment (bbelderbos) [inline characters/npc/npc_settings.json:26]: update images?

File: mechanics/input_processing.py [hunk 6]
```diff
@@ -1,48 +1,103 @@
...
+#             "Humor": 66,
+#             "Ambition": 66,
+#             "Affection": 0,
+#             "Confidence": 23,
+#             "Adventurousness": 87,
+#             "Empathy": 25
+
+
+def generate_response(npc_personality: dict, user_input: str) -> str:
+    message_sentiment = marvin.classify(
```
Comment (bbelderbos) [inline mechanics/input_processing.py:29]: so this part works exactly like textblob before right?

File: mechanics/input_processing.py [hunk 7]
```diff
@@ -1,48 +1,103 @@
...
+        user_input, labels=["friendly", "unfriendly", "neutral"]
+    )
+    message_type = marvin.classify(
+        user_input, labels=["greeting", "farewell", "request", "question", "other"]
+    )
+    top_values = [
+        element[0]
+        for element in (
+            sorted(npc_personality.items(), key=lambda item: item[1], reverse=True)[:2]
```
Comment (bbelderbos) [inline mechanics/input_processing.py:38]: why :2 ?

File: mechanics/input_processing.py [hunk 8]
```diff
@@ -1,48 +1,103 @@
...
+        )
+    ]
+    message_values = json.loads(
+        marvin.cast(
+            user_input,
+            target=str,
+            instructions="Get how much Intelligence, Honesty, Patience, Humor, Ambition, Affection, Confidence, Adventurousness or Empathy the message has; classify in a scale of 0-100",
```
Comment (bbelderbos) [inline mechanics/input_processing.py:45]: so here you tried `MessagePersonality` as a more specialized dict as return value? that should work?

File: mechanics/input_processing.py [hunk 9]
```diff
@@ -1,48 +1,103 @@
...
+        if higher_values == 0:
+            sentiment = "boring"
+        if higher_values == 1:
+            sentiment = "neutral"
+        if higher_values == 2:
+            sentiment = "excited"
+    else:
+        return response_semantics(npc_values, msg_type, user_input, sentiment)
+
+    return response_semantics(npc_values, msg_type, user_input, sentiment)
```
Comment (bbelderbos) [inline mechanics/input_processing.py:93]: seems these 2 returns are duplicated, can we ditch the one after else?

File: mechanics/input_processing.py [hunk 10]
```diff
@@ -1,48 +1,103 @@
...
+    sentiment: str,
+    msg_type: str,
+    npc_personality: dict,
+    npc_values: list,
+    message_values: dict,
+    user_input: str,
+) -> str:
+    higher_values = 0
+    if sentiment != "neutral" and msg_type == "other":
+        for value in npc_values:
```
Comment (bbelderbos) [inline mechanics/input_processing.py:81]: can you explain these 3 lines? not 100% clear to me

File: mechanics/input_processing.py [hunk 11]
```diff
@@ -1,48 +1,103 @@
...
+        user_input,
+    )
+    return response
+
+
+def response_semantics(
+    npc_values: list, msg_type: str, user_input: str, sentiment: str
+):
+    specifics = ", ".join(npc_values)
+    instructions = f"Respond this {msg_type} with a {sentiment} message highlighting the following values: {specifics} without giving away the value names"
```
Comment (bbelderbos) [inline mechanics/input_processing.py:64]: this is cool

File: mechanics/input_processing.py [hunk 12]
```diff
@@ -1,48 +1,103 @@
...
+        top_values,
+        message_values,
+        user_input,
+    )
+    return response
+
+
+def response_semantics(
+    npc_values: list, msg_type: str, user_input: str, sentiment: str
+):
```
Comment (bbelderbos) [inline mechanics/input_processing.py:62]: return hint

**Review by bbelderbos — APPROVED** [review]
This is cool, I wonder if we can simplify a bit more but these seem reusable functions so let me know how integration with the game logic goes ...

File: characters/npc/npc_settings.json [hunk 13]
```diff
@@ -1,36 +1,44 @@
...
-        "Humor": 86,
-        "Ambition": 18,
-        "Affection": 15,
-        "Confidence": 77,
-        "Adventurousness": 12,
-        "Empathy": 15
+    "npc1": {
+        "settings": {
+            "name": "npc1",
+            "image_path": "/Users/0xabaddon/new_repos/project_G/images/heart.png",
```
[the remark this answers is not in the document]
Comment (abaddonpuff) [reply characters/npc/npc_settings.json:5]: Its actually using relative paths! The image path will be generated from the function in main_game.py using pathlib. Just need to run the game once.

File: mechanics/input_processing.py (hunk 2 above)
Comment (abaddonpuff) [reply mechanics/input_processing.py:89]: cant be more than 2 because I only check the 2 top values :D

Comment (bbelderbos) [reply mechanics/input_processing.py:89]: but that relies on a hardcoded value elsewhere? what if that changes?

Comment (bbelderbos) [reply mechanics/input_processing.py:89]: is that this one? https://github.com/abaddonpuff/project_G/pull/18#discussion_r1698821795

**Pull request merged**